Extend FileIOFactory to Improve Customization

XJDKC opened the pull request:
# Context

Currently, we have three places where a `FileIO` is constructed:

1. `PolarisBaseCatalog::loadFileIO`: Used as the default `FileIO` across the entire catalog during testing.
2. `PolarisBaseCatalog::refreshIOWithCredential`: Refreshes storage credentials and constructs a new `FileIO`.
3. `TaskFileIOSupplier::apply`: Fetches subscoped credentials and uses `FileIOFactory` to create a `FileIO` for cleanup tasks.

We’ve identified an issue where not all properties required by `FileIO` are generated by the `FileIOFactory`. For instance, subscoped credentials are retrieved outside the `FileIOFactory` and then passed to it. Additionally, the current `FileIOFactory` interface has limited access to contextual information.

To enhance the ability to customize `FileIOFactory`, we plan to extend it by passing additional information. This improvement will enable developers to implement custom logic such as auditing, proxy, and more.

Some code pointers:
* [PolarisBaseEntity.java#L61-L62](https://github.com/apache/polaris/blob/5aa76fba723550e48af38d3e6fa916560b7d9336/polaris-core/src/main/java/org/apache/polaris/core/entity/PolarisBaseEntity.java#L61-L62)
* [TaskFileIOSupplier.java#L56-L88](https://github.com/apache/polaris/blob/main/service/common/src/main/java/org/apache/polaris/service/task/TaskFileIOSupplier.java#L56-L88)

<!--
    Possible security vulnerabilities: STOP here and contact security@apache.org instead!

    Please update the title of the PR with a meaningful message - do not leave it "empty" or "generated"
    Please update this summary field:

    The summary should cover these topics, if applicable:
    * the motivation for the change
    * a description of the status quo, for example the current behavior
    * the desired behavior
    * etc

    PR checklist:
    - Do a self-review of your code before opening a pull request
    - Make sure that there's good test coverage for the changes included in this PR
    - Run tests locally before pushing a PR (./gradlew check)
    - Code should have comments where applicable. Particularly hard-to-understand
      areas deserve good in-line documentation.
    - Include changes and enhancements to the documentation (in site/content/in-dev/unreleased)
    - For Work In Progress Pull Requests, please use the Draft PR feature.

    Make sure to add the information BELOW this comment.
    Everything in this comment will NOT be added to the PR description.
-->

snazy commented:
Thanks for your contribution @XJDKC!
I think this change blurs the line between "properties" and "internal properties" quite a bit. I'm also, in general, concerned about generic property bag per se containing all kinds of information - things that can be exposed (upstream/downstream) and things that must not - nobody can ensure that no "bad things" (think: logging secrets) happen to this information.

dimas-b commented on service/common/src/main/java/org/apache/polaris/service/catalog/BasePolarisCatalog.java:
```diff
@@ -1627,6 +1627,12 @@ private FileIO refreshIOWithCredentials(
+
+    // Populate the internal properties to FileIO in case the FileIO implementation needs them
+    Map<String, String> internalProperties =
+        storageInfoEntity.map(PolarisEntity::getInternalPropertiesAsMap).orElse(Map.of());
```
I agree with @snazy 's concern (from the main comment thread) about the distinction between "properties" and "internal properties".

Currently, this is not well-defined in the code (e.g. there's no javadoc explaining what is "internal"), so some clarification is probably called for.

That said, I think supporting customized access to storage is a valid use case. We just need to pay more attention to defining the model for it so as to avoid confusion in the future if/when related code need to be refactored.

@collado-mike : WDYT (as the original code contributor)?

XJDKC replied:
Thank you for sharing your concerns! Here’s some clarification and context regarding this change from my POV:

1. **Distinction Between `properties` and `internalProperties`**:
    * `properties` is the public property map that will be exposed to the REST Client.
      * Reference: [IcebergCatalogAdapter::getConfig](https://github.com/apache/polaris/blob/5320af566a8829e39a55b59d6d309e7c326ad5ae/service/common/src/main/java/org/apache/polaris/service/catalog/IcebergCatalogAdapter.java#L607-L650)
    * `internalProperties` is strictly internal to Polaris and is never propagated to the REST Client. This ensures that sensitive or implementation-specific details remain isolated.
      * For example, the storage configuration map is serialized as a JSON string and stored in internalProperties.
2. **Mitigation Against "Bad Things"**:
    * The info in the `internalProperties` is not logged or exposed externally, as it is handled within Polaris’s controlled environment.

Since this PR only propagates internalProperties to Polaris’s FileIO (and not to the REST Client), it enables Polaris’s FileIO to implement custom logic without impacting client-facing behavior. So I think it should be fine.

cc: @snazy @dimas-b @collado-mike @eric-maynard @dennishuo

snazy replied:
Maybe bikeshedding, but there is no "Polaris FileIO" - it's running code from a different project, that can change at any time.

IMHO the issue is rather to have properly typed configurations rather than two generic property bags, where everything can end in. Sure, property bags are easy to start with - but in my experience always cause issues later on - and it's nearly impossible to fix that "later".

It looks that this change is meant to deal with credentials, which should IMHO be handled with a lot of care.

sfc-gh-rxing replied:
> there is no "Polaris FileIO"

The `Polaris FileIO` here means the FileIO used by Polaris to access the storage, for aws s3, it's `S3FileIO`, for Azure Storage, it's `ADLSFileIO`. It could also be a custom FileIO since we can inject FileIOFactory

> Is rather to have have properly typed configurations rather than two generic property bags

But in the current Iceberg SDK and Polaris pattern, customers can add any properties to the `properties` map in PolarisEntity.

> this change is meant to deal with credentials

This change is not meant to deal with credentials, **credentials are already passed to FileIO without this change**. This change is trying to pass some existing storage config to FileIO. Also this change doesn't add any new property in the `internalProperties` map, for now, `internalProperties` map contains the following info:
```json
{
  "storage_configuratio_info": "{
    "@type" : "AwsStorageConfigurationInfo",
    "storageType" : "S3",
    "allowedLocations" : [ "s3://externally-owned-bucket", "s3://my-bucket/path/to/data" ],
    "roleARN" : "arn:aws:iam::012345678901:role/jdoe",
    "fileIoImplClassName" : "org.apache.iceberg.aws.s3.S3FileIO",
    "externalId" : "externalId"
  }"
}
```

XJDKC replied:
> there is no "Polaris FileIO"

The `Polaris FileIO` here means the FileIO used by Polaris to access the storage, for aws s3, it's `S3FileIO`, for Azure Storage, it's `ADLSFileIO`. It could also be a custom FileIO since we can inject FileIOFactory

> Is rather to have have properly typed configurations rather than two generic property bags

But in the current Iceberg SDK and Polaris pattern, customers can add any properties to the `properties` map (the public map, not the `internalProperties`) in PolarisEntity.

> this change is meant to deal with credentials

This change is not meant to deal with credentials, **credentials are already passed to FileIO without this change** (Code Pointer: [BasePolarisCatalog.java#L1589-L1619](https://github.com/apache/polaris/blob/734e6979ccd010109323544c9f6dc9f0dcaecb8b/service/common/src/main/java/org/apache/polaris/service/catalog/BasePolarisCatalog.java#L1589-L1619
)). This change is trying to pass some existing storage config to FileIO. Also this change doesn't add any new property in the `internalProperties` map, for now, `internalProperties` map contains the following info:
```json
{
  "storage_configuratio_info": "{
    "@type" : "AwsStorageConfigurationInfo",
    "storageType" : "S3",
    "allowedLocations" : [ "s3://externally-owned-bucket", "s3://my-bucket/path/to/data" ],
    "roleARN" : "arn:aws:iam::012345678901:role/jdoe",
    "fileIoImplClassName" : "org.apache.iceberg.aws.s3.S3FileIO",
    "externalId" : "externalId"
  }"
}
```

sfc-gh-rxing replied:
How about introducing a configuration called `ALLOWED_FILE_IO_PROPERTIES`. If the property is included in the allowlist, we pass it to FileIO. Otherwise, we just filter it. 

Pseudocode:
```java
public void refreshIOWithCredentials(...) {
    // ......
    Map<String, String> additionalProperties = new HashMap<>();
    storageInfoEntity.getInternalProperties().forEach((key, value) -> {
        if (ALLOWED_FILE_IO_PROPERTIES.contains(key)) {
            additionalProperties.put(key, value);
        }
    });
    tableProperties.putAll(additionalProperties);
    FileIO fileIO = null;
    fileIO = loadFileIO(ioImplClassName, tableProperties);
    // ensure the new fileIO is closed when the catalog is closed
    closeableGroup.addCloseable(fileIO);
    return fileIO;
}
```

collado-mike replied:
Passing in the entire set of table internal properties definitely seems wrong to me. For example, in the case of federated catalog items, one can imagine some api key or password to access a remote Iceberg catalog might be found in internal properties. I wouldn't want to pass that to the `FileIO` with the assumption that it's simply going to ignore what it doesn't understand.

sfc-gh-rxing replied:
Thanks for the reply, that's a good point. Given this, would the allowlist proposal be acceptable?

People who add properties to the allowlist should take responsibility for handling them properly and securely.

XJDKC replied:
Thanks for the reply, that's a good point. Given this, would the allowlist proposal be acceptable?

People who add properties to the allowlist should take responsibility for handling them properly and securely.

dennishuo replied:
The intent of `internalProperties` really is to have room for some Polaris-internal persistence items that don't *directly* interface with any kind of external code, but rather are contained by some Polaris-specific interpretation layer. So, I'm not sure an allowlist is enough because it still mixes the domain of config keys that show up in `internalProperties` with the domain of config keys that happen to be Iceberg-recognizable properties.

Indeed, some internalProperties are translated into "top-level" structured fields in REST objects, others only control internal behaviors, etc.

So, my thinking here is we should probably be explicit about what we're trying to do. Specifically, we are trying to give *additional* internal contextual information to a `FileIOFactory.loadFileIO` call (a Polaris construct) *not* necessarily to an underlying `FileIO` (an Iceberg core construct). We might already be improperly mixing too much into just the open-ended properties map we give to the factory, but IMO we should update the `FileIOFactory` interface to:

    FileIO loadFileIO(String impl, Map<String, String> properties, PolarisResolvedPathWrapper targetEntity);

Maybe in the future even explicitly passing in some form of call context

    FileIO loadFileIO(String impl, Map<String, String> properties, PolarisResolvedPathWrapper targetEntity, CallContext callContext);

The use case here is for better injectable functionality in the form of the `FileIOFactory` for Polaris service-owners where custom logic that might relate to custom internalProperties can easily intercept the construction of new FileIO instances with all the necessary information about the current request available.

dennishuo replied:
I guess my suggestion is basically just apossible formulation of what @dimas-b was talking about -- adding a translation layer even if we might happen to use a "trivial" translation in some early use cases. So, +1 to having that layer -- in this case we can make it clear that the `FileIOFactory` extension provided by custom service-owners is what is responsible for the interpretation here.

sfc-gh-rxing replied:
Passing the `PolarisResolvedPathWrapper` / `CallContext` to FileIOFactory is a great suggestion!

Compared to passing the allowed internalProperties, it brings two benefits:
* **`FileIOFactory` as the Translation Layer**:
  * For the allowed properties solution, `refreshIOWithCredentials` only performs basic translation by filtering internalProperties, which is very inflexible.
  * Moving the translation logic to FileIOFactory enables more complex and flexible handling of the info.
* **Bringing All Necessary Information While Preventing Key Conflicts**:
  * By passing the `PolarisResolvedPathWrapper` and `CallContext` to `FileIOFactory`, it ensures all necessary information is passed and available for processing, enhancing flexibility and helping to avoid potential key conflicts.

Since FileIOFactory is a Polaris interface, Polaris has full control over its handling. For FileIOFactories in OSS Polaris, OSS community should proper process of the passed objects. While for custom `FileIOFactory`, handling is the responsibility of the Polaris customizer.

XJDKC replied:
Passing the `PolarisResolvedPathWrapper` / `CallContext` to FileIOFactory is a great suggestion!

Compared to passing the allowed internalProperties, it brings two benefits:
* **`FileIOFactory` as the Translation Layer**:
  * For the allowed properties solution, `refreshIOWithCredentials` only performs basic translation by filtering internalProperties, which is very inflexible.
  * Moving the translation logic to FileIOFactory enables more complex and flexible handling of the info.
* **Bringing All Necessary Information While Preventing Key Conflicts**:
  * By passing the `PolarisResolvedPathWrapper` and `CallContext` to `FileIOFactory`, it ensures all necessary information is passed and available for processing, enhancing flexibility and helping to avoid potential key conflicts.

Since FileIOFactory is a Polaris interface, Polaris has full control over its handling. For FileIOFactories in OSS Polaris, OSS community should proper process of the passed objects. While for custom `FileIOFactory`, handling is the responsibility of the Polaris customizer.

XJDKC replied:
Another similar approach could be to store the `PolarisResolvedPathWrapper` in `CallContext::contextVariables`. By doing so, we might avoid the need to pass it explicitly to FileIOFactory.

Since contextVariables already contains the catalog instance, it seems like a natural fit for encapsulating additional contextual data like `PolarisResolvedPathWrapper`.

sfc-gh-rxing replied:
Hey @snazy @dimas-b WDYT about the new solution?
If you feel okay with passing `PolarisResolvedPathWrapper` to `FileIOFactory`, I will go for it!

XJDKC replied:
Hey @snazy @dimas-b WDYT about the new solution?
If you feel okay with passing PolarisResolvedPathWrapper to FileIOFactory, I will go for it!

dimas-b replied:
TBH, I do not see much difference in passing a `Map` vs. `PolarisResolvedPathWrapper` to `FileIOFactory` because both approaches use essentially type-less containers of generic properties and the meaning of those properties is _implicit_ and subject to runtime / build variations.

Would it be possible to have something like `FileIOFactory.loadFileIO(PolarisCatalogConfig config)` where `PolarisCatalogConfig` would have a `type` method returing `S3`, `GCS`, etc... plus typed "credentials", etc.?

The idea being that `PolarisCatalogConfig` is strongly typed in Polaris core and persisted. `FileIOFactory` implementations use it to make Iceberg FileIO property bags in runtime. When the server is upgraded, the _same_ `PolarisCatalogConfig` is loaded, but factory implementations may be _different_ and initialize FileIO objects differently. However, `FileIOFactory` implementations still follow the contract of `PolarisCatalogConfig`. OSS implementations would use only the strongly typed part of `PolarisCatalogConfig` for operation.

If a particular `FileIOFactory` implementation needs some config that is not yet defined in Polaris core, and does not "fit" in OSS, the config would use the `type` of `MY-CUSTOM-S3` (as an example) and provide for extra generic properties. Ideally, the parsing of config properties (on loading from persistence) is also delegated to a pluggable ckass that knows how to deal with `MY-CUSTOM-S3`. WDYT?

If that is too much, for this PR, but you agree in principle, how about removing `String impl` from `FileIOFactory.loadFileIO` and adding a well-know type property plus javadoc that the factory is responsible for interpreting the `type`. Then, we could pass both "normal" and "internal" properties to `FileIOFactory.loadFileIO` as a short-term solution.

```
public interface FileIOFactory {
  // type is specifically NOT the Iceberg FileIO class name
  FileIO loadFileIO(String type, Map<String, String> properties, Map<String, String> internalProperties);
}
```

eric-maynard replied:
On the contrary, I actually think having the FileIO take unstructured data aligns well with our current model and best supports pluggability. 

Polaris (Iceberg) table properties are a `Map<String, String>`, and FileIO's `properties` returns a `Map<String, String>`. Forcing this map through some type like `PolarisCatalogConfig` seems unnecessary and limiting. 

If someone wants to bring their own FileIO implementation and configure some table to interact with it, I would prefer if Polaris could support that without any code changes needed.

XJDKC replied:
Thanks for sharing your thoughts!

To inject dynamic properties from persistence layer to `BasePolarisCatalog`, the dynamic properties will be attached to the entity object. So PolarisEntity it serves as the data source, while `FileIO` acts as the data consumer. A translation layer between them is indeed necessary.

Regarding the `PolarisCatalogConfig` solution, it does involve some translation in `BasePolarisCatalog`, where we need to derive a `PolarisCatalogConfig` instance from the `PolarisEntity` object. However, this approach seems limiting. For instance, while we can infer the `type` from `PolarisStorageConfigurationInfo`, it doesn’t provide a flexible way to handle custom types like `MY-CUSTOM-S3`. If the translation logic is fixed in BasePolarisCatalog, we lose the flexibility to adapt to varying types dynamically. i.e. How could we pass `MY-CUSTOM-S3` to `FileIOFactory` with a fixed translation logic in OSS Polaris.

By shifting the translation logic to `FileIOFactory`, we centralize and streamline it. This ensures that the `FileIOFactory` can directly interpret the dynamic properties and handle custom logic, such as supporting `MY-CUSTOM-S3`, without relying on a predefined translation mechanism in `BasePolarisCatalog`.

Additionally, even if we pass a general type like `S3` to `FileIOFactory`, it can still interpret dynamic properties to handle specific cases like `MY-CUSTOM-S3`. This ensures extensibility and keeps BasePolarisCatalog more generic.

For the short term solution, I think we can just still pass `impl` to `FileIOFactory` while passing additional `internalProperties`. `impl` is a kind of `type` from `FileIOFactory` perspective. But I'm okay with either options.

collado-mike replied:
Rather than a new `PolarisCatalogConfig` type, can we reuse the `PolarisStorageConfigurationInfo`? Seems like that's an existing type that should probably contain all of the information needed to configure the FileIO, with the obvious exception of credentials. Adding custom properties to that type actually makes a lot of sense so that users can configure connection details when they create it.

sfc-gh-rxing replied:
Yes, I would prefer using `PolarisStorageConfigurationInfo`. 

Also, considering the point Eric mentioned, users may override the FileIO impl, for example, they may want Polaris to use `HadoopFileIO` to access the storage: see [BasePolarisCatalog.java#L239-L257](https://github.com/apache/polaris/blob/be8cc359ce17eeab1f66a175092c8291b6703992/service/common/src/main/java/org/apache/polaris/service/catalog/BasePolarisCatalog.java#L239-L257). So passing `ioImplClassName` to FileIOFactory is necessary.

Since `PolarisStorageConfigurationInfo` is part of `internalProperties`, we can pass `internalProperties` directly:
```
public interface FileIOFactory {
  FileIO loadFileIO(String impl, Map<String, String> properties, Map<String, String> internalProperties);
}
```

XJDKC replied:
Yes, I would prefer using `PolarisStorageConfigurationInfo`. 

Also, considering the point Eric mentioned, users may override the FileIO impl, for example, they may want Polaris to use `HadoopFileIO` to access the storage: see [BasePolarisCatalog.java#L239-L257](https://github.com/apache/polaris/blob/be8cc359ce17eeab1f66a175092c8291b6703992/service/common/src/main/java/org/apache/polaris/service/catalog/BasePolarisCatalog.java#L239-L257). So passing `ioImplClassName` to FileIOFactory is necessary.

Since `PolarisStorageConfigurationInfo` is part of `internalProperties`, we can pass `internalProperties` directly:
```java
public interface FileIOFactory {
  FileIO loadFileIO(String impl, Map<String, String> properties, Map<String, String> internalProperties);
}

public class CustomFileIOFactory {
  FileIO loadFileIO(String impl, Map<String, String> properties, Map<String, String> internalProperties) {
    String proxyEndpoint = internalProperties.get(PROXY_ENDPOINT);
    properties.add(S3FileIOProperties.ENDPOINT, proxyEndpoint); // translation logic
    return CatalogUtil.loadFileIO(impl, properties, new Configuration());
  }
}
```

collado-mike replied:
No, I wouldn't pass in the the `internalProperties` directly for all the reasons already mentioned. The `BasePolarisCatalog` already fetches the `PolarisStorageConfigurationInfo` for the table before it loads the `FileIO`. We can pass the storage info there.

dennishuo replied:
I still think we should just pass in the PolarisResolvedPathWrapper, which is the superset of all the other suggested fields in a way that doesn't prematurely mix or cherrypick fields from the entity.

1. Even if *today* the only "interesting" field in the resolved entity is `internalProperties` that doesn't always have to be true. If we pass the resolved entity, then in the future if we introduce additional structured fields, it doesn't change what we're plumbing into the factory
2. The semantic here is to make it so that BasePolarisCatalog isn't trying to cherrypick which "interesting" properties we think a factory might need. Give the factory all the info to make a decision. Even if we decided to extract specifically the `PolarisStorageConfigurationInfo` that's also trying to state that we think the `PolarisStorageConfigurationInfo` is the only thing the factory might need, which isn't true
3. My other reason for suggesting `PolarisResolvedPathWrapper` is that we actually then get the entity hierarchy as well in the factory. There are various reasons why custom factories might need the full hierarchy.

dennishuo replied:
Another way to look at this is to consider the fact that `BasePolarisCatalog::refreshIOWithCredentials` is already trying to serve as a consolidated factory method for a FileIO. And the method signature here indicates that we think the following information is all pertinent to the construction of a FileIO:

      TableIdentifier identifier,
      Set<String> readLocations,
      PolarisResolvedPathWrapper resolvedStorageEntity,
      Map<String, String> tableProperties,
      Set<PolarisStorageActions> storageActions

We only later added a FileIOFactory, because some custom users of Polaris didn't want to fork the code and modify the body of `BasePolarisCatalog::refreshIOWithCredentials`. So in a sense the purpose of `FileIOFactory` is to say a custom Polaris deployment is capable of fully customizing the way a FileIO is constructed, possibly with layers of e.g. throttling, auditing, whatever without ever needing to "fork" Polaris code, but only needing to configure an injected `FileIOFactory` class.

In this sense, there's an argument here for plumbing *all* those arguments through to the `FileIOFactory`.

If the factory wants to emit an alert whenever `readLocations` contains the substring `secret-passwords`, then it can easily do that. If the factory wants to redirect all requests that have `WRITE` actions through a special auditing proxy, it can do that too. If it wants to look at the `entityVersion` and emit special logs whenever the entityVersion is divisible by 42, it can also do that.

XJDKC replied:
Agree with Dennis, we should move the logic in `refreshIOWithCredentials` to `FileIOFactory` so that `FileIOFactory` has the full control on how to construct a FileIO and become a real **factory**.

collado-mike replied:
I think moving the credentials generation into the `FileIOFactory` makes a ton of sense. And, IMO, makes for a much more cohesive and sensible design than simply passing in the the table properties. I'd be in favor of changing the signature of `FileIOFactory` to take in all of the params necessary to generate credentials, then custom implementations can also do whatever they want with table properties.

I would change that signature, though, to take in the whole `TableLikeEntity`, rather than just the `tableProperties`. It's unclear from that signature whether the externally exposed properties field is being used or if it's the `internalProperties`. I can imagine scenarios where both/either are needed by the factory.

XJDKC replied:
To summarize all the ideas, here is the proposed new `FileIOFactory` definition:
```java
public interface FileIOFactory {
  FileIO loadFileIO(
      String impl,
      TableIdentifier identifier,
      Set<String> readLocations,
      PolarisResolvedPathWrapper resolvedStorageEntity,
      Map<String, String> tableProperties,
      Set<PolarisStorageActions> storageActions);
}
```

cc: @snazy @dimas-b WDYT?

dimas-b replied:
What is the purpose of `resolvedStorageEntity` in this API?

XJDKC replied:
As mentioned before, we will move logic from `refreshIOWithCredentials` to `FileIOFactory`, so `FileIOFactory` is responsible to get all the required info for constructing a FileIO. This includes getting the subcoped creds. So we need `resolvedStorageEntity`.

dimas-b replied:
To clarify my concerns from previous message, let's consider this use case:
1. An OSS (admin) user deploys Polaris version 1.0.
2. The user configures access to storage. This step involved defining some Iceberg FileIO properties, naturally.
3. There's a need/request to refactor Polaris internals in v. 2.0
4. The user upgrades Polaris to 2.0
5. How do we make sure the user's configuration still works? Note that we do not know what the user actually configured.

IMHO, at step 5 we are bound by having to support Iceberg FileIO implementations and deal with changes in their behaviour. My point is that in order for Polaris to be able carry out this duty effectively, the configuration has to be explicit. That is to say, people evolving the Polaris code have to know exactly what properties are configurable by end users. For example, with an generic property bag approach, it will be very hard to use non-Iceberg FileIO in Polaris 2.0 (for example) even for its _interrnal_ access to storage.

XJDKC replied:
For Step 2:
The only way a user can configure access to storage is by defining `table-default.` properties at the catalog level. These properties are included in the `tableProperties` argument of `refreshIOWithCredentials` and passed to FileIO.

Since we are passing all the information that `refreshIOWithCredentials` has to `FileIOFactory`, this ensures that `FileIOFactory` also receives the tableProperties and can pass them to `FileIO`. Therefore, the case you mentioned does not pose any issues.

Regarding the generic property bag approach: users can already set arbitrary properties using the `table-default.` prefix. This functionality inherently creates a generic property bag, which is outside the scope of this PR.

The new proposal is trying to move the logic of `refreshIOWithCredentials` to `FileIOFactory` which doesn't break existing feature.

dennishuo replied:
We could probably benefit from some terminology to distinguish between:

1. End-user who calls Polaris REST APIs
2. "Polaris codebase/project User" who is an infrastructure admin who takes apache/polaris and deploys it into a production environment for other "end-users" to interact with
    2a. "Default" Polaris users who usually just take a jarfile from maven central and run it with minimal customizations to configuration
    2b. "Customized" Polaris users who mix in lots of proprietary plugins and may have a fork of the core Polaris code

If I understand correctly, the main concern is about how the Apache Polaris project itself can ensure "project users" (2) can be supported as the code evolves.

@XJDKC 's point seems relevant though that at the lowest common denominator, there's already a property bag that can be set by the end-users (1) that can already plumb through behavioral changes either in the base Iceberg FileIO directly or by interpretation from the FileIOFactory, regardless of this PR. Whether we should make a change to lock that down more and/or enumerate "supported" properties could be worth a separate discussion outside of this PR.

This intersects with concerns about refactors of Polaris internals because in theory (2b) users could privately come to rely on injection of custom-meaning properties into that public `properties` map as much as they could do so on `internalProperties`.

As I mentioned before, IMO passing in the `resolvedStorageEntity` doesn't preclude adding more rigid "structure" to what is being used in the factory and in particular it needn't encourage any direct use of `internalProperties` either.

In particular, we can still make the `DefaultFileIOFactory` effectively enumerate the officially supported set of configuration that we'll be careful as a project to ensure carries over from 1.0 to 2.0, etc. The "Default" polaris service-runners (2a) will only be using such out-of-the-box factories, so we'll be able to know exactly what we're allowed to evolve without breaking (2a).

Those in the (2b) category are explicitly signing up to do their own legwork *if* they both add plumbing of custom internal properties and use those custom properties in a custom FileIOFactory, and I think what we're saying here is that this extensibility mechanism is an explicit design choice to allow such advanced use cases.

I suppose alternatively, as a project we could assert that we want to avoid encouraging such extensibility mechanisms, and that a persistent code-fork is the preferred way to customize FileIO construction in such ways instead.

Both approaches kind of say the same thing -- "deep customizations mean you have to do your own homework on major-version backwards compatibility", and I guess requiring a code fork is a bit more explicit, but I guess my main point is that we can still convey that delineation of a backwards-compatibility contract despite allowing an "easier" extensibility mechanism.

dimas-b replied:
Re: the "common denominator" point, I understand that there may be existing features in Polaris what allow for wide and unrestricted customizations. However, I do not think it is a valid reason to justify exposing _new_ areas for customization without imposing some structure.

For example, currently, storage configuration saved in the internal properties has to map to `PolarisStorageConfigurationInfo` (and sub-classes). So, inputs to `FileIOFactory` as "sanitized" by going through those classes. With this PR, it will be possible for admin users to configure storage access parameters that are not reflected in Polaris code, and as such are obscure from the Polaris developers' POV.

It is not my intention to block this PR, but I still do not see how my example (with upgrades) can work conveniently for OSS users in the wild and for Polaris developers. We can always say that whoever puts custom values into "internal" properties will be in the "unsupported" upgrade land, but that, IMHO, it too much of a burden for users. Not all users will want or be able to or deal with java code internals in Polaris / Iceberg (by extension).

To restate my previous suggestions in more concrete terms, I think it would be beneficial for the customizations that this PR is expected to enable to be first-class properties in `PolarisStorageConfigurationInfo` (or sub-classes).

Then, the conversion of internal properties to FileIO properties can be done by `FileIOFactory` taking a `PolarisStorageConfigurationInfo` as input, or by Polaris Core (passing a generic `Map` to `FileIOFactory`). From my POV this concern is secondary if we establish a clear set of what is possible to configure.

Again, if the expected customizations in the internal properties are not suitable for general OSS use, I think we ought to introduce another plugin point (e.g. `FileIOConfigurer`) with the default OSS impl. working the same way as `refreshIOWithCredentials()`. This way it is simply not possible for OSS user to enter the unsupported land without manually deploying their own Polaris extension code.

XJDKC replied:
> With this PR, it will be possible for admin users to configure storage access parameters that are not reflected in Polaris code, and as such are obscure from the Polaris developers' POV.

Why would OSS Polaris developers need to know the custom properties configured by users of customized Polaris setups? These properties are meant for user-specific customization and should not require visibility or concern from OSS developers unless they directly impact OSS functionality.

> I still do not see how my example (with upgrades) can work conveniently for OSS users in the wild and for Polaris developers.

For OSS users, the logic of refreshIOWithCredentials is moved to DefaultFileIOFactory. If users do not customize the FileIO, they can upgrade directly without issues.

For "Customized" Polaris users, if they customize Polaris by injecting properties into internalProperties, they can still upgrade directly unless internalProperties is completely removed.
Additionally, since all information available to `refreshIOWithCredentials` is passed to `FileIOFactory` (including resolved entities), custom properties can be stored elsewhere in the entity object. This is one of the reasons why we avoid passing `internalProperties` directly to `FileIOFactory` and instead pass all arguments from `refreshIOWithCredentials`.

> This PR is expected to enable to be first-class properties in PolarisStorageConfigurationInfo

How can a custom `PolarisMetastoreManager` inject a property not defined in the OSS `PolarisStorageConfigurationInfo` as the first class property?

> Introduce another plugin point (e.g. FileIOConfigurer)
```java
public class FileIOConfig {
  private final String impl;
  private final Map<String, String> properties;
}
public interface FileIOConfigurer {
  FileIOConfig generateConfig(
      String impl,
      TableIdentifier identifier,
      Set<String> readLocations,
      PolarisResolvedPathWrapper resolvedStorageEntity,
      Map<String, String> tableProperties,
      Set<PolarisStorageActions> storageActions);
}

public void refreshIOWithCredentials(...) {
    // ......
    FileIOConfigurer fileIOConfigurer = new DefaultFileIOConfiguer(...);
    FileIOConfig fileIOConfig = fileIOConfigurer.generateConfig();
    FileIO fileIO = loadFileIO(FileIOConfig);
    // .....
    return fileIO;
}

public class FileIOFactory {
  FileIO loadFileIO(FileIOConfig) {
    String proxyEndpoint = internalProperties.get(PROXY_ENDPOINT);
    properties.add(S3FileIOProperties.ENDPOINT, proxyEndpoint); // translation logic
    // Only pass properties to FileIO
    return CatalogUtil.loadFileIO(impl, properties, new Configuration());
  }
}
```

This is the proposed solution?  From my POV, this solution is almost the same as the the new `FileIOFactory` proposal but we split the logic into two classes.

dimas-b replied:
> Why would OSS Polaris developers need to know the custom properties configured by "Customized" Polaris users?

Anyone can deploy Polaris OSS code and effectively be the admin of that deployment. We cannot know beforehand what those users will do. My concerns are about how the Polaris developers can deal (hopefully without a huge burden) with supporting those deployments.

dimas-b replied:
> For OSS users, the logic of refreshIOWithCredentials is moved to DefaultFileIOFactory. If users do not customize the FileIO, they can upgrade directly without issues.

I did not see that in this PR... Did you push that change?

dennishuo replied:
I believe this was just based on what's being discussed, perhaps @XJDKC didn't want to push the change until we got to agreement on the comments here.

It sounds like we're actually in violent agreement though :)

From my comment earlier:

> In particular, we can still make the DefaultFileIOFactory effectively enumerate the officially supported set of configuration that we'll be careful as a project to ensure carries over from 1.0 to 2.0, etc. The "Default" polaris service-runners (2a) will only be using such out-of-the-box factories, so we'll be able to know exactly what we're allowed to evolve without breaking (2a).
> Those in the (2b) category are explicitly signing up to do their own legwork if they both add plumbing of custom internal properties and use those custom properties in a custom FileIOFactory, and I think what we're saying here is that this extensibility mechanism is an explicit design choice to allow such advanced use cases.

So whether we expose 1 or 2 plugin classes, the basic idea is that the `FileIOFactory` at least already represents the main plugin entrypoint, and it will effectively define exactly the fields used when constructing a FileIO (whether or not it offloads this to an additional "`FileIOConfiguerer`" as you suggest), and anyone using Polaris OSS will by default use the well-supported DefaultFileIOFactory.

Only those who write additional custom code that uses custom properties will go "off-road" so to speak, That's the (2b) category.

XJDKC replied:
Hey @dimas-b, I have summited a commit with the latest proposal. It's not a perfect commit cuz I only update the DefaultFileIOFactory (I can't even compile it). Hopefully it will help!

dimas-b commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/FileIOFactory.java:
```diff
@@ -19,9 +19,21 @@
   FileIO loadFileIO(String impl, Map<String, String> properties);
+
+  FileIO loadFileIO(
+      String impl,
```
nit: maybe `className`?

dimas-b commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/DefaultFileIOFactory.java:
```diff
@@ -19,18 +19,157 @@
+
+  private final RealmContext realmContext;
+  private final PolarisEntityManager entityManager;
+  private final PolarisMetaStoreManager metaStoreManager;
```
Should this be `PolarisCredentialVendor`? It seems to be the narrowest interface in use on this object in this class.

XJDKC replied:
Yes, we can narrow down the scope.

dimas-b commented on service/common/src/main/java/org/apache/polaris/service/catalog/BasePolarisCatalog.java:
```diff
@@ -1592,21 +1592,38 @@ private FileIO refreshIOWithCredentials(
+    //        storageInfoEntity
+    //            .map(
+    //                storageInfo ->
+    //                    refreshCredentials(identifier, storageActions, readLocations,
```
Why keep this code commented out?

XJDKC replied:
Removed!

dimas-b commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/FileIOFactory.java:
```diff
@@ -19,9 +19,21 @@
 
 /** Interface for providing a way to construct FileIO objects, such as for reading/writing S3. */
 public interface FileIOFactory {
   FileIO loadFileIO(String impl, Map<String, String> properties);
```
Is this method still needed?

XJDKC replied:
The new changes are only used to showcase the overall idea, so I just keep both. If we agree on this proposal, I will just remove the old interface and start to work on a complete implementation.

dimas-b replied:
As for me, the new `loadFileIO` API and the default impl is quite reasonable. Let's clean this PR up and open it general review :+1:

XJDKC replied:
Should I open another PR or just keep using this one?

dimas-b replied:
I'd keep using this PR. Since there were no other reviews so far, I think it should be ok to squash existing commits, rebase and force push. Then the PR will be as new :)

dimas-b reviewed (commented):
Thanks for the updated @XJDKC ! Overall, the new code looks good to me :) A few minor comments below.

snazy commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/DefaultFileIOFactory.java:
```diff
@@ -19,18 +19,157 @@
+
+  @Override
+  public FileIO loadFileIO(
+      String impl,
```
Not related to this PR, but this is a real annoyance in Iceberg.
There can be symbolic FileIO implementation "names" and class names - and neither practically works across all programming languages (Iceberg/Java has its special implementations, PyIceberg has its own, ...).

snazy replied:
Anyways, maybe rename this parameter to something like "icebergFileIoIdentifier"

snazy replied:
WFM - just wanted to note that behavior ;)

snazy commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/DefaultFileIOFactory.java:
```diff
@@ -19,18 +19,157 @@
+    return storageInfoEntity;
+  }
+
+  private Map<String, String> refreshCredentials(
```
Actually, since this PR changes FileIO/object-storage-credentials stuff, I think it would be good to separate and expose functionality from the interface (or a separate one?) to enable support for Iceberg's refresh-credentials via its REST API.

XJDKC replied:
Create a separate util class so that both the `BasePolarisCatalog` and `FileIOFactory` can use the same methods.
For `BasePolarisCatalog`, we need to generate/refresh subscoped creds when the client sends a loadTable request.
For `DefaultFileIOFactory`, we need to get subscoped creds first so that we can access the storage and read/write metadata json files.

snazy commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/DefaultFileIOFactory.java:
```diff
@@ -19,18 +19,157 @@
+  private static @Nonnull Optional<PolarisEntity> findStorageInfoFromHierarchy(
+      PolarisResolvedPathWrapper resolvedStorageEntity) {
+    Optional<PolarisEntity> storageInfoEntity =
+        resolvedStorageEntity.getRawFullPath().reversed().stream()
```
maybe add a note that this starts at the "leaf" (table) and walks "upwards" via namespaces to the "root"

snazy commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/DefaultFileIOFactory.java:
```diff
@@ -19,18 +19,157 @@
+                || storageActions.contains(PolarisStorageActions.ALL)
+            ? tableLocations
+            : Set.of();
+    Map<String, String> credentialsMap =
```
absolute nitpick and totally my own personal preference: using Java's (new-ish) `var` keyword actually increases readability, assuming that the variable name is expressive.

snazy commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/DefaultFileIOFactory.java:
```diff
@@ -19,18 +19,157 @@
+  }
+
+  private PolarisCredentialVendor getCredentialVendor() {
+    return metaStoreManager;
```
Better narrow the type and change the name of this field - and inline it's usage.

snazy commented on quarkus/service/src/test/java/org/apache/polaris/service/quarkus/catalog/BasePolarisCatalogTest.java:
```diff
@@ -1459,6 +1462,46 @@ public void testDropTableWithPurgeDisabled() {
   }
 
+  @Test
+  public void testRefreshIOWithCredentialsAndInternalProperties() {
```
Not sure how this actually _refreshes_ credentials.

Anyway, having this test is good. However, I'd prefer to also have a dedicated test class for `DefaultFileIOFactory` and do extensive testing there, also with multiple realms?

snazy commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/FileIOFactory.java:
```diff
@@ -19,9 +19,21 @@
+import org.apache.polaris.core.persistence.PolarisResolvedPathWrapper;
+import org.apache.polaris.core.storage.PolarisStorageActions;
 
 /** Interface for providing a way to construct FileIO objects, such as for reading/writing S3. */
```
Can you add a comment here saying something like
```
 *
 * <p>Implementations are available via CDI as {@link ApplicationScoped @ApplicationScoped} beans.
```

sfc-gh-rxing commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/DefaultFileIOFactory.java:
```diff
@@ -19,18 +19,157 @@
 public class DefaultFileIOFactory implements FileIOFactory {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DefaultFileIOFactory.class);
+
+  private final RealmContext realmContext;
```
We need to get configuration based on the current realm. So I mark `DefaultFileIOFactory` as `@RequestScoped` instead.

XJDKC replied:
We need to get configuration based on the current realm. So I mark DefaultFileIOFactory as @RequestScoped instead.

XJDKC commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/DefaultFileIOFactory.java:
```diff
@@ -19,18 +19,157 @@
+    return fileIO;
+  }
+
+  private static @Nonnull Optional<PolarisEntity> findStorageInfoFromHierarchy(
```
Hey @snazy, thank you for reviewing! The previous version was just a demo to illustrate the idea, so I copied what we had in `refreshIOWithCredential` over to `DefaultFileIOFactory`. Very appreciate your detailed feedback, I’ll revise accordingly!

snazy commented on quarkus/service/src/test/java/org/apache/polaris/service/quarkus/catalog/BasePolarisCatalogTest.java:
```diff
@@ -1458,6 +1461,55 @@ public void testDropTableWithPurgeDisabled() {
+                PolarisConfiguration.SKIP_CREDENTIAL_SUBSCOPING_INDIRECTION.key,
+                false));
+    try (MockedStatic<CatalogUtil> catalogUtil =
+        Mockito.mockStatic(CatalogUtil.class, Mockito.CALLS_REAL_METHODS)) {
```
Oh, please do not mock anything static. I know this is not powermock, but mocking static stuff requires some "interesting" instrumentation. You wrap the call to CatalogUtil from DefaultFileIOFactory in a separate method and override that method here. Something like:

```java
class DefaultFileIOFactory {
   @VisibleForTesting
   FileIO callCatalogUtil(params) {
     return CatalogUtil.load...
   }
}
```
and here 
```java
new DefaultFileIOFactory(params) {
   @Override 
   FileIO callCatalogUtil(params) {
     // run your assertions
     return super.callCatalogUtil(...)   // or not :shrug:
}
```

XJDKC replied:
Done, create a separate class for the FileIOFactory.

snazy commented on service/common/src/main/java/org/apache/polaris/service/catalog/BasePolarisCatalog.java:
```diff
@@ -2040,7 +2034,13 @@ private List<TableIdentifier> listTableLike(PolarisEntitySubType subType, Namesp
+    return fileIOFactory.loadFileIO(
+        ioImpl,
+        propertiesWithS3CustomizedClientFactory,
+        null /*tableIdentifier*/,
+        null /*tableLocations*/,
+        null /*storageActions*/,
+        null /*resolvedStorageEntity*/);
```
`null` looks a bit dangerous to me. Maybe add a `loadFileIO(ioImpl,properties)` and let the implementation do the "right" thing?

snazy commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/FileIOFactory.java:
```diff
@@ -19,9 +19,19 @@
 public interface FileIOFactory {
-  FileIO loadFileIO(String impl, Map<String, String> properties);
+  FileIO loadFileIO(
+      String ioImplClassName,
+      Map<String, String> properties,
+      TableIdentifier identifier,
+      Set<String> tableLocations,
+      Set<PolarisStorageActions> storageActions,
+      PolarisResolvedPathWrapper resolvedStorageEntity);
```
Can you add `@Nonnull` to the parameters? (see also my other comment about `null`)

snazy commented on service/common/src/main/java/org/apache/polaris/service/task/TaskFileIOSupplier.java:
```diff
@@ -83,6 +83,6 @@ public FileIO apply(TaskEntity task, RealmId realmId) {
         properties.getOrDefault(
             CatalogProperties.FILE_IO_IMPL, "org.apache.iceberg.io.ResolvingFileIO");
-    return fileIOFactory.loadFileIO(ioImpl, properties);
+    return fileIOFactory.loadFileIO(ioImpl, properties, null, null, null, null);
```
Same comment about `null` as above

snazy commented on gradle/libs.versions.toml:
```diff
@@ -76,6 +76,7 @@ junit-bom = { module = "org.junit:junit-bom", version = "5.11.4" }
 logback-classic = { module = "ch.qos.logback:logback-classic", version = "1.5.16" }
 micrometer-bom = { module = "io.micrometer:micrometer-bom", version = "1.14.3" }
 mockito-core = { module = "org.mockito:mockito-core", version = "5.15.2" }
+mockito-inline = { module = "org.mockito:mockito-inline", version = "5.2.0" }
```
Hm. This one looks quite unmaintained - last release was 2023. Do we really need it?

XJDKC commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/DefaultFileIOFactory.java:
```diff
@@ -19,18 +19,157 @@
+  private static final Logger LOGGER = LoggerFactory.getLogger(DefaultFileIOFactory.class);
+
+  private final RealmContext realmContext;
+  private final PolarisEntityManager entityManager;
```
It's used when getting the subscoped creds.

XJDKC commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/DefaultFileIOFactory.java:
```diff
@@ -19,18 +19,157 @@
+    return metaStoreManager;
+  }
+
+  private Boolean getBooleanContextConfiguration(String configKey, boolean defaultValue) {
```
Inline the function instead!

XJDKC commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/DefaultFileIOFactory.java:
```diff
@@ -19,18 +19,157 @@
+      PolarisEntityManager entityManager,
+      PolarisMetaStoreManager metaStoreManager,
+      PolarisMetaStoreSession metaStoreSession,
+      PolarisConfigurationStore polarisConfigurationStore) {
```
Thanks for pointing it out! The previous version was just a demo to illustrate the idea. Now I pushed the code with the actual implementation, this class is annotated as RequestScoped now!

snazy commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/DefaultFileIOFactory.java:
```diff
@@ -19,18 +19,157 @@
 /** A simple FileIOFactory implementation that defers all the work to the Iceberg SDK */
 @ApplicationScoped
 @Identifier("default")
 public class DefaultFileIOFactory implements FileIOFactory {
```
Do not extend any bean class - especially not if you pass other beans to the parent class (which is the case here), even if that's `abstract` class. It effective makes such beans unproxyable - Quarkus fails at build time - Weld, the CDI reference implementation fails at runtime.

snazy replied:
Similar for functions that return or receive beans. Adding `final` to those functions doesn't help. `final` alone makes a bean unproxyable.

dennishuo reviewed (approved):
LGTM

sfc-gh-dhuo commented:
@snazy @dimas-b Looks like all the major comments have been resolved; any final/remaining suggestions before merging this?

XJDKC commented:
Hey @snazy @dimas-b May I get another review for this PR, thanks!

dimas-b commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/FileIOFactory.java:
```diff
@@ -18,10 +18,50 @@
+/**
+ * Interface for providing a way to construct FileIO objects, such as for reading/writing S3.
+ *
+ * <p>Implementations are available via CDI as {@link ApplicationScoped @ApplicationScoped} beans.
```
nit: why mention `@ApplicationScoped` on the interface that does not carry this annotation? IMHO, this is not an API concern, but an integration/build concern.

XJDKC replied:
Suggested by @snazy in this comment: https://github.com/apache/polaris/pull/724#discussion_r1924753312

dimas-b replied:
I'm fine to keep this comment :)

dimas-b commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/FileIOFactory.java:
```diff
@@ -18,10 +18,50 @@
 public interface FileIOFactory {
-  FileIO loadFileIO(String impl, Map<String, String> properties);
+
+  /** This method is intended for use in tests only. */
```
... but it is called from `BasePolarisCatalog`. I think we ought to refactor that call path to avoid having extension point APIs that have to provide different functionality in test contexts (can be done later).

XJDKC replied:
Done!

dimas-b commented on service/common/src/main/java/org/apache/polaris/service/catalog/BasePolarisCatalog.java:
```diff
@@ -1586,27 +1525,22 @@ protected String viewName() {
   }
 
-  private FileIO refreshIOWithCredentials(
+  private FileIO refreshIOForTableLike(
```
nit: why call it "refresh FileIO" it does not "refresh" any FileIO objects but makes a new one.

XJDKC replied:
Yes, I think `refresh` here means we refresh the credential used by the FileIO. But instead of updating the existing FileIO, we create a new one instead since there is no such a way to updating the FileIO's properties.

I will rename the method to `loadFileIOForTableLike` instead.

dimas-b commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/WasbTranslatingFileIOFactory.java:
```diff
@@ -19,19 +19,63 @@
+      MetaStoreManagerFactory metaStoreManagerFactory,
+      PolarisConfigurationStore configurationStore) {
+    defaultFileIOFactory =
+        new DefaultFileIOFactory(
```
How is the behaviour of `WasbTranslatingFileIOFactory` different from `DefaultFileIOFactory`? Why do we need `WasbTranslatingFileIOFactory`?

XJDKC replied:
When we load the FileIO, we wrap the FileIO loaded by `DefaultFileIOFactory` with `WasbTranslatingFileIO`. `WasbTranslatingFileIO` will traslate the `wasb` file path to `abfs` file path.

dimas-b replied:
I guess I missed `WasbTranslatingFileIO` :sweat_smile:  IMHO, wrapping an implementation that has a specific purpose and behaviour (like `DefaultFileIOFactory`) is not ideal. I'd prefer to extract a base class that can be shared between `DefaultFileIOFactory` and `WasbTranslatingFileIOFactory`... but that can be done later.

XJDKC replied:
Will do it in a separate PR! Another dumb question, is it the best practice to extend a abstract for an ApplicationScoped bean considering this? https://github.com/apache/polaris/pull/724#discussion_r1925795624

eric-maynard replied:
@dimas-b  we didn't take that approach because `WasbTranslatingFileIO` was/is supposed to be very temporary until the Iceberg FileIO implementations support WASB. This was originally merged, then reverted, so I'm not sure on the current status. But the goal should remain to make `WasbTranslatingFileIO` easy to rip out.

dimas-b replied:
@XJDKC : I believe @snazy 's concern was about extending classes that are themselves CDI beans (i.e. have CDI annotations). I do not think there's any problem with extending simple java classes and putting CDI annotations on the child class.

dimas-b replied:
@eric-maynard : like I commented we can address this later... If we can remove `WasbTranslatingFileIO` at that point that will remove my concern too :)

XJDKC replied:
I see, maybe we can keep this as it's since the long term plan is to remove it.

dimas-b commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/DefaultFileIOFactory.java:
```diff
@@ -18,19 +18,112 @@
+                        storageInfo))
+            .orElse(Map.of());
+
+    // Update the FileIO before we write the new metadata file
```
What are we updating here?.. and why mention "metadata file"?

XJDKC replied:
I just copied this comment from `refreshIOWithCredentials`, will revise the comment.

XJDKC replied:
Done!

dimas-b commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/FileIOUtil.java:
```diff
@@ -0,0 +1,129 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FileIOUtil {
```
nit: for a `public` util class with only static methods, it is best to add a `private` constructor to prevent accidental instantiation.

XJDKC replied:
Good catch! I will add a `private` constructor instead!

dimas-b commented on service/common/src/test/java/org/apache/polaris/service/catalog/io/FileIOFactoryTest.java:
```diff
@@ -0,0 +1,228 @@
+            Set.of(storageLocation),
+            Set.of(PolarisStorageActions.READ),
+            resolvedPathWrapper)
+        .close();
```
nit: `assertThatCode(...).doesNotThrowAnyException()`

dimas-b commented on quarkus/service/src/test/java/org/apache/polaris/service/quarkus/catalog/io/TestFileIOFactory.java:
```diff
@@ -18,23 +18,32 @@
  */
 @Vetoed
-public class TestFileIOFactory extends DefaultFileIOFactory {
+public class TestFileIOFactory implements FileIOFactory {
```
nit: `TestFileIOFactory` -> `ObservingFileIOFactory` or `MeasuredFileIOFactory`...

`TestFileIOFactory` is confusing wrt. `FileIOFactoryTest`.

sfc-gh-rxing replied:
Done!

dimas-b reviewed (commented):
Sorry about the delay. The latest changes LGTM with one remaining concern about `WasbTranslatingFileIOFactory` (which I'd like to clarity) and a few nit comments (which are optional).

snazy commented:
Recent changes LGTM - deferring to @dimas-b

dimas-b reviewed (commented):
I believe @XJDKC mentioned making a few minor follow-up changes... otherwise LGTM.

sfc-gh-rxing commented on service/common/src/main/java/org/apache/polaris/service/catalog/BasePolarisCatalog.java:
```diff
@@ -2039,8 +1975,7 @@ private List<TableIdentifier> listTableLike(PolarisEntitySubType subType, Namesp
   private FileIO loadFileIO(String ioImpl, Map<String, String> properties) {
-    Map<String, String> propertiesWithS3CustomizedClientFactory = new HashMap<>(properties);
-    return fileIOFactory.loadFileIO(ioImpl, propertiesWithS3CustomizedClientFactory);
+    return CatalogUtil.loadFileIO(ioImpl, properties, new Configuration());
```
Thanks for reviewing this PR! I also refactored the TestServices:
1. Move it to polaris-service-common module
2. Allow us to provide some custom objects when creating the services.
Hopefully it can be helpful for our testing.

dimas-b replied:
The current state of this PR looks acceptable to me. I still this we should avoid the call to `CatalogUtil.loadFileIO(...)` in `BasePolarisCatalog` and always use the factory. Ideally, we should not have test-specific logic in production code, such as that controlled by the `INITIALIZE_DEFAULT_CATALOG_FILEIO_FOR_TEST` property in `BasePolarisCatalog`... but that can be done later.

XJDKC replied:
Used `FileIOFactory` instead to load the catalog default FileIO.

XJDKC commented:
Hi folks, I think this PR is ready to be merged into main branch.
I have addressed all the comments, there are still some followup tasks, but we can do them later: 
e.g. remove test only `loadFileIO` interface from `PolarisBaseCatalog`.

Also I have updated the PR desc to clarify the purposes for this PR. 
Can someone help merging it into main branch if there is no other issue? Thanks!
cc: @dimas-b @snazy @dennishuo @eric-maynard @collado-mike

eric-maynard commented on service/common/src/main/java/org/apache/polaris/service/catalog/io/FileIOUtil.java:
```diff
@@ -0,0 +1,131 @@
+      LOGGER
+          .atInfo()
+          .addKeyValue("tableIdentifier", tableIdentifier)
+          .log("Skipping generation of subscoped creds for table");
```
This might be too verbose at `INFO`

XJDKC replied:
Done!
